samples/android/app/src/main/java/org/hyperledger/indy/sdk/ui/ledger/LedgerActivity.kt

```diff
@@ -0,0 +1,314 @@
+package org.hyperledger.indy.sdk.ui.ledger
+
+import android.os.Bundle
+import android.util.Log
+import android.view.View
+import androidx.appcompat.app.AppCompatActivity
+import kotlinx.android.synthetic.main.activity_anoncreds.*
+import kotlinx.android.synthetic.main.activity_endorser.*
+import kotlinx.android.synthetic.main.activity_ledger.*
+import kotlinx.coroutines.*
+import org.hyperledger.indy.sdk.R
+import org.hyperledger.indy.sdk.did.Did
+import org.hyperledger.indy.sdk.did.DidJSONParameters
+import org.hyperledger.indy.sdk.helpers.MessageHelper
+import org.hyperledger.indy.sdk.ledger.Ledger
+import org.hyperledger.indy.sdk.pool.Pool
+import org.hyperledger.indy.sdk.utils.PoolUtils
+import org.hyperledger.indy.sdk.wallet.Wallet
+import org.json.JSONObject
+import org.junit.Assert
+import java.lang.Exception
+
+class LedgerActivity : AppCompatActivity() {
+
+    private val TAG = LedgerActivity::class.java.name
+
+
+    // my vars
+    private val trusteeSeed = "000000000000000000000000Trustee1"
+    private lateinit var myWallet: Wallet
+    private lateinit var trusteeWallet: Wallet
+    private lateinit var trusteeDid: String
+    private lateinit var myDid: String
+    private lateinit var myVerkey: String
+    private lateinit var pool: Pool
+    private lateinit var nymRequest: String
+    private lateinit var myWalletConfig: String
+    private lateinit var myWalletCredentials: String
+    private lateinit var trusteeWalletConfig: String
+    private lateinit var trusteeWalletCredentials: String
+    private lateinit var poolName: String
+
+
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+        setContentView(R.layout.activity_ledger)
+
+        // start Ledger demo
+        startDemo()
+    }
+
+    private fun updateUI(text: String) {
+        tvLedgerLogs.text = "${tvLedgerLogs.text}$text"
+    }
+
+    private fun updateHeader(text: String) {
+        pbLedger.visibility = View.VISIBLE
+        tvLedgerStart.text = text
+    }
+
+    private fun updateFooter(text: String) {
+        pbLedger.visibility = View.GONE
+        tvLedgerEnd.text = text
+    }
+
+    /**
+     * startDemo function start all functions for Ledger demo chronological in coroutine default thread
+     */
+    private fun startDemo() {
+        MainScope().launch {
+            Log.d(TAG, "startDemo: Ledger sample -> STARTED!")
+
+            updateHeader(getString(R.string.ledger_sample_start))
+
+
+            updateUI(getString(R.string.ledger_create_ledger))
```

Comment:
You have a lot of duplicates here. Essentially you are doing the same thing for each step. 
1. write message
2. perform an action in background
3. handle error
4. write status

this should be extracted in an helper function that you should pas just parameters in with messages and action to perform.

Reply:
also you are doing this in multiple activities. try to reuse logic across activities.